README.md
# Hull Webhooks Ship.

The Outgoing Webhooks Connector sends a payload for an entity (user or account) if an event or change has been registered. For the user 
entity, payloads can be sent on event occurrences, attribute changes, and/or segment changes. For the account entity, payloads 
can be sent on attribute changes and/or segment changes. Both entities require that a global filter of segments
be defined. This list is used to filter out any user or account that does not belong to a listed segment. If the user
or account belongs to a defined segment, any changes or events on the entity will be checked as configured to determine 
if a payload should be sent. However, there is the option to send the payload regardless of an entity event occurrence 
and/or segment change given that the entity belongs to a defined segment.

If you want your own instance: [![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy?template=https://github.com/hull-ships/hull-outgoing-webhooks)

---

### Using :

[See Readme here](https://dashboard.hullapp.io/readme?url=https://hull-outgoing-webhooks.herokuapp.com)

### Developing :

- Fork
- Install

```sh
npm install
npm run start:dev
```

### Logs :

  Below list represents all specific logs for Webhooks Connector :

    * notification.error - logged when sending webhook operation returned some errors


CHANGELOG.md
# Changelog

## v0.3.10
- added support for kraken filtering

## v0.3.9
- updated status to take away warnings that were not bad

## v0.3.8
- use kraken exports for batches

## v0.3.7
- support account batches

## v0.3.6
- removing multiple channels in manifest

## v0.3.5
- removed a faulty required field validation in manifest

## v0.3.4
- account integration support

## v0.3.3
- upgrade hull-node to overcame batch endpoint failure

## v0.3.2
- improve documentation
- add `account` when processing batch extracts

## v0.3.1
- add `message_id` param to `outgoing.user.start` log line

## v0.3.0
- remove archetype builder
- upgrade dependencies, inlcuding hull-node@0.13.16
- test are run by jest now
- new standard file `notifications-configuration.js` introduced

## v0.2.16
- specify subscription conditions

## v0.2.15
- upgrade hull-node to 0.13.15

## v0.2.14
- add flow control env vars

## v0.2.13
- upgrade hull-node to 0.13.14
- changed http client to not follow redirects and treat them as errors

## v0.2.12
- adds `throttle_rate` and `throttle_per_rate` settings to control throughput
- refactor code into SyncAgent

## v0.2.11
- update documentation for concurrency

## v0.2.10
- treat 200-204 status codes as success

## v0.2.9
- add concurrency support
- upgrade hull-node
- cleanup babeljs and flow configuration, rewrite import/export syntax
- replace request with superagent

## v0.2.8

- re-add smart-notifier
- use hull-connector archetype
- add status endpoint
- refine logging

## v0.2.7

- revert smart-notifier support

## v0.2.6

- add more details to error logs

## v0.2.5

- enable to send all users
- updated pictures

## v0.2.4

- update to hull-node@0.11.8
- update logging to match convention

## v0.2.3

- update to hull-node@0.11.4
- move files to lib directory

## v0.2.2

- update to hull-node@0.11.0
- implement logging convention

## v0.2.1

- update to hull-node@beta
- add batch metrics

## v0.2.0

- [BREAKING] changing the name of segments filter (`segment_filter` -> `synchronized_segments`)
- add prefixing selected attributes with `traits_` to enhance matching (sometimes the event comes with the prefix while the attribute key in settings it without it)
- introduce segments based filtering on the `/batch` endpoint
- enhance the documentation in readme and in settings UI
- cleanup the code
- replace `utils` with `hull-ship-base` prototype

## v0.1.0

- introduce error handling and metrics


server/lib/webhook.js
/* @flow */
const _ = require("lodash");
const superagent = require("superagent");
const Promise = require("bluebird");
const {
  superagentInstrumentationPlugin,
  superagentErrorPlugin
} = require("hull/lib/utils");

const { version } = require("../../package.json");

function webhook(
  { ctx, url, hull, payload = {}, metric }: Object,
  throttle: Object,
  targetEntity: "user" | "account"
) {
  if (!_.isNil(url)) {
    url = url.trim();
  }
  const maxNumRetries = process.env.MAX_RETRIES || 5;
  const numRetries = _.get(ctx, "notification.kraken.retries", 0);
  const smartNotifierResponse = _.get(ctx, "smartNotifierResponse");
  let asTargetEntity;
  if (targetEntity === "user") {
    asTargetEntity = hull.asUser(
      _.pick(payload.user, ["id", "email", "external_id"])
    );
  } else if (targetEntity === "account") {
    asTargetEntity = hull.asAccount(_.pick(payload.account, ["id", "domain"]));
  }
  let start;
  return superagent
    .post(url)
    .set("X-Hull-Outgoing-Webhooks", `v${version}`)
    .on("request", () => {
      start = process.hrtime();
    })
    .use(throttle.plugin())
    .use(superagentErrorPlugin({ timeout: 20000 }))
    .use(
      superagentInstrumentationPlugin({
        logger: hull.logger,
        metric
      })
    )
    .redirects(0)
    .ok(res => res.status >= 200 && res.status <= 204)
    .send(payload)
    .then(response => {
      const status = response.status;
      const hrTime = process.hrtime(start);
      const elapsed = hrTime[0] * 1000 + hrTime[1] / 1000000;
      asTargetEntity.logger.info(`outgoing.${targetEntity}.success`, {
        status,
        url,
        elapsed
      });
      asTargetEntity.logger.debug("webhook.success", {
        payload,
        response: response.body
      });
      return elapsed;
    })
    .catch(error => {
      const errorInfo = {
        reason: "Webhook Failed",
        status: error.status,
        error: error.message,
        message: "See error param for further details about the exact error."
      };

      const retryWebhook =
        numRetries < maxNumRetries &&
        (error.status === 429 || error.status >= 500);

      if (retryWebhook) {
        // smartNotifierResponse could be nil if we are consuming a Batch.
        if (smartNotifierResponse && smartNotifierResponse.setFlowControl) {
          smartNotifierResponse.setFlowControl({
            type: "retry",
            in: (error.response.headers["Retry-After"] || 120) * 1000
          });
        }
      }
      const res = {
        payload,
        hull_summary: "",
        error: errorInfo,
        status: error.status
      };
      if (error.status >= 300 && error.status <= 308) {
        res.hull_summary =
          "Server returned a redirect code - see `Connector doesn't support redirects` in documentation";
      }
      asTargetEntity.logger.error(`outgoing.${targetEntity}.error`, res);
      return Promise.resolve(res);
    });
}

module.exports = webhook;
